onnxruntime/core/providers/xnnpack/nn/matmul.cc

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "matmul.h"
+#include "core/graph/constants.h"
+#include "core/graph/graph.h"
+#include "core/graph/graph_utils.h"
+#include "core/framework/transpose_helper.h"
+#include "core/providers/utils.h"
+#include "core/providers/xnnpack/detail/utils.h"
+#include "core/framework/tensorprotoutils.h"
+#include "core/providers/cpu/math/matmul_helper.h"
+
+namespace onnxruntime {
+namespace xnnpack {
+
+bool MatMul::IsMatMulOnnxNodeSupported(const NodeUnit& node_unit, const GraphViewer& graph) {
+
+  bool supported = false;
+  const onnxruntime::Node& node = node_unit.GetNode();
+
+  // use do {} while(false) so it's easier to set a breakpoint on the return
+  do {
+    auto input_defs = node.InputDefs();
+
+    if (input_defs.size() != 2) {
+      break;
+    }
+
+    const auto& A_arg = *input_defs[0];
+    const auto& B_arg = *input_defs[1];
+
+    // Support only float
+    const auto* A_type = A_arg.TypeAsProto();
+    const auto* B_type = B_arg.TypeAsProto();
+
+    if (A_type == nullptr || B_type == nullptr ||
+        A_type->tensor_type().elem_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT ||
+        B_type->tensor_type().elem_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT ) {
+        break;
+    }
+
+    // B matrix must be constant
+    if (B_arg.Exists() && graph.GetConstantInitializer(B_arg.Name(), true) == nullptr) {
+        break;
+    }
+
+    // making sure we are dealing with MatMul
+    const auto* B_shape = B_arg.Shape();
+
+    if (!B_shape || B_shape->dim_size() > 3) {
```

Comment:
Can you explain this check? I thought the MatMulComputeHelper handled a range of different ranks and limiting this to 2 or 3D wouldn't be necessary. Naively if anything I would expect it to be limited to 2D though.